// react-native-css-interop/dist/runtime/native/inheritance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createColorSchemeSignal = exports.InheritanceProvider = exports.effectContext = exports.globalVariables = void 0;
const react_1 = require("react");
const signals_1 = require("../signals");
const color_scheme_1 = require("./color-scheme");
exports.globalVariables = {
    root: new Map(),
    universal: new Map(),
};
exports.effectContext = (0, react_1.createContext)({
    signals: exports.globalVariables.root,
});
exports.InheritanceProvider = exports.effectContext.Provider;
function createColorSchemeSignal(id) {
    let light = (0, signals_1.createSignal)(undefined, `${id}#light`);
    let dark = (0, signals_1.createSignal)(undefined, `${id}#dark-app`);
    const get = () => {
        return color_scheme_1.colorScheme.get() === "light"
            ? light.get()
            : dark.get() ?? light.get();
    };
    const peek = () => {
        return color_scheme_1.colorScheme.peek() === "light"
            ? light.peek()
            : dark.peek() ?? light.peek();
    };
    const unsubscribe = (subscription) => {
        dark.unsubscribe(subscription);
        light.unsubscribe(subscription);
    };
    const set = (value) => {
        if ("dark" in value)
            dark.set(value.dark);
        if ("light" in value)
            light.set(value.light);
    };
    return {
        id,
        get,
        set,
        peek,
        unsubscribe,
    };
}
exports.createColorSchemeSignal = createColorSchemeSignal;
