// react-native-css-interop/dist/testing-library/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerCSS = exports.resetStyles = exports.createRemappedComponent = exports.createMockComponent = exports.warnings = exports.styleMetaMap = exports.globalStyles = void 0;
const react_1 = require("react");
const JSX = __importStar(require("react/jsx-runtime"));
const index_1 = require("../index");
const render_1 = require("../runtime/render");
const shared_1 = require("../shared");
const css_to_rn_1 = require("../css-to-rn");
var misc_1 = require("../runtime/native/misc");
Object.defineProperty(exports, "globalStyles", { enumerable: true, get: function () { return misc_1.globalStyles; } });
Object.defineProperty(exports, "styleMetaMap", { enumerable: true, get: function () { return misc_1.styleMetaMap; } });
var stylesheet_1 = require("../runtime/native/stylesheet");
Object.defineProperty(exports, "warnings", { enumerable: true, get: function () { return stylesheet_1.warnings; } });
require("../core-components");
__exportStar(require("../types"), exports);
function createMockComponent(Component, { mapping = { className: "style" }, } = {}) {
    (0, index_1.cssInterop)(Component, mapping);
    const mock = jest.fn((props, _ref) => {
        return (0, render_1.render)(JSX.jsx, Component, props, "");
    });
    const component = (0, react_1.forwardRef)(mock);
    return Object.assign(component, { mock });
}
exports.createMockComponent = createMockComponent;
function createRemappedComponent(Component, mapping) {
    (0, index_1.remapProps)(Component, mapping);
    const mock = jest.fn((props, _ref) => {
        return (0, render_1.render)(JSX.jsx, Component, props, "");
    });
    const component = (0, react_1.forwardRef)(mock);
    return Object.assign(component, { mock });
}
exports.createRemappedComponent = createRemappedComponent;
exports.resetStyles = index_1.StyleSheet[shared_1.INTERNAL_RESET].bind(index_1.StyleSheet);
function registerCSS(css, options) {
    index_1.StyleSheet.register((0, css_to_rn_1.cssToReactNativeRuntime)(css, options));
}
exports.registerCSS = registerCSS;
